saleor/warehouse/models.py

```diff
@@ -0,0 +1,145 @@
+    id = models.UUIDField(default=uuid.uuid4, primary_key=True)
+    name = models.CharField(
+        pgettext_lazy("Warehouse field description", "Warehouse name"), max_length=255
+    )
+    company_name = models.CharField(
+        pgettext_lazy("Warehouse field description", "Legal company name"),
+        blank=True,
+        max_length=255,
+    )
+
+    shipping_zones = models.ManyToManyField(ShippingZone, blank=True)
+    address = models.ForeignKey(Address, on_delete=models.PROTECT)
+
+    email = models.EmailField(
+        pgettext_lazy("Warehouse field description", "Email address"),
+        blank=True,
+        default="",
+    )
+
+    objects = WarehouseQueryset.as_manager()
+
+    class Meta:
+        ordering = ("-name",)
+        permissions = (
+            (
+                "manage_warehouses",
+                pgettext_lazy("Permission description", "Manage warehouses."),
+            ),
+        )
+
+    def __str__(self):
+        return self.name
+
+    @property
+    def countries(self) -> Set[str]:
+        shipping_zones = self.shipping_zones.all()
+        return set(itertools.chain(*[zone.countries for zone in shipping_zones]))
+
+    def delete(self, *args, **kwargs):
+        address = self.address
+        super().delete(*args, **kwargs)
+        address.delete()
+
+
+class StockQuerySet(models.QuerySet):
+    def annotate_available_quantity(self):
+        return self.annotate(available_quantity=F("quantity") - F("quantity_allocated"))
+
+    def for_country(self, country_code: str):
+        query_warehouse = models.Subquery(
+            Warehouse.objects.filter(
+                shipping_zones__countries__contains=country_code
+            ).values("pk")
+        )
+        return self.select_related("product_variant", "warehouse").filter(
+            warehouse__in=query_warehouse
+        )
+
+    def get_variant_stock_for_country(
+        self, country_code: str, product_variant: ProductVariant
+    ):
+        return self.for_country(country_code).get(product_variant=product_variant)
+
+
+class Stock(models.Model):
+    warehouse = models.ForeignKey(Warehouse, null=False, on_delete=models.PROTECT)
+    product_variant = models.ForeignKey(
+        ProductVariant, null=False, on_delete=models.CASCADE
+    )
+    quantity = models.PositiveIntegerField(default=0)
+    quantity_allocated = models.PositiveIntegerField(default=0)
+
+    objects = StockQuerySet.as_manager()
+
+    class Meta:
+        unique_together = [["warehouse", "product_variant"]]
+        permissions = (
+            (
+                "manage_stocks",
+                pgettext_lazy("Permission description", "Manage stocks."),
+            ),
+        )
+
+    def __str__(self):
+        return f"{self.product_variant} - {self.warehouse.name}"
+
+    @property
+    def quantity_available(self) -> int:
+        return max(self.quantity - self.quantity_allocated, 0)
+
+    @property
+    def is_available(self):
```

Comment:
What do you think about renaming it to `is_in_stock`? We used to have a function like this and it was checking if the quantity available is above zero so that would be similar to the previous implementation. Also, we tend to describe a product as available if it is in stock and it's published.

Reply:
We have two functions added - `is_variant_in_stock` and `is_product_in_stock` that should be used in other places (they use `stock.is_available`). I'm open to suggestion to rename this property but not exactly for `is_in_stock`. If we call it as `stock.is_in_stock` then this sounds funny ;)

Reply:
I don't have any better idea.. maybe it can stay as it is- it's not that big deal.